common/numeric-limits.h

```diff
@@ -0,0 +1,89 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef COMMON_NUMERIC_LIMITS_H
+#define COMMON_NUMERIC_LIMITS_H
+
+#include "common/scummsys.h"
+#include "common/type-traits.h"
+
+namespace Common {
+
+template <typename T, bool TIsIntegral, bool TIsUnsigned>
+struct IntegralNumericLimits {
+};
+
+template<typename T>
+struct IntegralNumericLimits<T, true, false> {
+	static const T max_value = ((T(1) << (sizeof(T) * 8 - 2)) - 1) * 2 + 1;
```

Comment:
I am not sure this formula is safe to use.
From where does it come?
I would rather inspire from what LLVM does: https://github.com/llvm/llvm-project/blob/main/libcxx/include/limits#L213

Reply:
This formula should be safe.  Basically we want a bit-fill of 1 less than the number of bits in the int, but we can't just do e.g. `(1<<31)-1` for int32, because 0x80000000-1 is an arithmetic overflow, which is undefined behavior.  So instead it computes a bit-fill of 2 bits less, e.g. `(1<<30)-1` for int32, then inserts a low 1 bit by multiplying it by 2 and adding 1.

Unsigned ints are the same story but I think it might not actually be necessary there because (unlike signed ints) unsigned ints are defined as having modulo behavior on wraparound so max can just be -1.

Reply:
OK, this should be safe